leader-election-impl/src/main/java/com/palantir/leader/LeadershipObserver.java

```diff
@@ -0,0 +1,25 @@
+/*
+ * Copyright 2018 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.leader;
+
+public interface LeadershipObserver {
+    void gainedLeadership();
+    void lostLeadership();
+
+    void executeWhenGainedLeadership(Runnable task);
+    void executeWhenLostLeadership(Runnable task);
```

Comment:
I think for this class in general, we should be more explicit (javadoc?) about what guarantees we are giving the user especially if there are multiple runnables submitted. We can be explicit that we are not guaranteeing much (for example, even though right now it looks like the callbacks are executed in order sequentially, we should be explicit as to whether we allow users of this to assume this behaviour.)

Reply:
Added javadoc 👍